src/libmongoc/src/mongoc/mongoc-generation-map.c

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2021-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "mongoc-generation-map-private.h"
+
+#include "utlist.h"
+
+typedef struct _gm_node_t {
+   bson_oid_t key;
+   uint32_t val;
+   struct _gm_node_t *next;
+} gm_node_t;
+
+static gm_node_t *
+gm_node_new ()
+{
+   return bson_malloc0 (sizeof (gm_node_t));
+}
+
+static void
+gm_node_destroy (gm_node_t *node)
+{
+   if (!node) {
+      return;
+   }
+   bson_free (node);
+}
+
+static gm_node_t *
+gm_node_copy (const gm_node_t *node)
+{
+   gm_node_t *node_copy = gm_node_new ();
+
+   bson_oid_copy (&node->key, &node_copy->key);
+   node_copy->val = node->val;
+   return node_copy;
+}
+
+struct _mongoc_generation_map {
+   gm_node_t *list;
+};
+
+mongoc_generation_map_t *
+mongoc_generation_map_new ()
+{
+   mongoc_generation_map_t *gm;
+
+   gm = bson_malloc0 (sizeof (mongoc_generation_map_t));
+   return gm;
+}
+
+mongoc_generation_map_t *
+mongoc_generation_map_copy (const mongoc_generation_map_t *gm)
+{
+   mongoc_generation_map_t *gm_copy;
+   gm_node_t *iter;
+
+   gm_copy = mongoc_generation_map_new ();
+
+   LL_FOREACH (gm->list, iter)
+   {
+      gm_node_t *node_copy;
+
+      node_copy = gm_node_copy (iter);
+      LL_PREPEND (gm_copy->list, node_copy);
+   }
+
+   return gm_copy;
+}
+
+uint32_t
+mongoc_generation_map_get (const mongoc_generation_map_t *gm,
+                           const bson_oid_t *key)
+{
+   gm_node_t *iter = NULL;
+
+   BSON_ASSERT (key);
+
+   LL_FOREACH (gm->list, iter)
+   {
+      if (bson_oid_equal (key, &iter->key)) {
+         break;
+      }
+   }
+
+   if (!iter) {
+      return 0;
+   }
+
+   return iter->val;
+}
+
+void
+mongoc_generation_map_increment (mongoc_generation_map_t *gm,
+                                 const bson_oid_t *key)
+{
+   gm_node_t *match;
+   gm_node_t *iter = NULL;
+
+   BSON_ASSERT (key);
```

Comment:
Since we don't appear to BSON_ASSERT(gm), I'll assume there's a reason we don't worry about it, and won't mention it further.